### ISSUE TITLE
Autocomplete: Multi-lang search (based on user's lang)

### ISSUE THREAD

**Issue by Joxit** [pelias/api#1296]
> Transcription of https://github.com/pelias/api/issues/127#issuecomment-490002821

## What is this for ?

We want Pelias to send responses to queries written in other languages than English.
For example, a Dutch looking for `Parijs` (Paris in Dutch) will get `Parijs, Frankrijk`.

## What should we do ?

- [ ] Use `multi_match` in autocomplete queries
- [ ] Should search in `name.$LANG` index with higher boost
- [ ] Should search in `name.default` index with standard boost
- [ ] Should search in `name.en` index as fallback with lower boost (when `$LANG` is not `en`)

## Some use cases 

| Text | Lang | Result |
| --- | --- | --- |
| Parijs | nl | Parijs, Frankrijk (whosonfirst:locality:101751119 Paris) |
| Londre | fr | Londres, Angleterre, Royaume-Uni (whosonfirst:locality:101750367 London) |
| ブラジル | ja | ブラジル (whosonfirst:country:85633009 Brazil) |

cc @mihneadb Have you started working on it ? I can take the task if you want :smile:

**Comment by mihneadb** [comment]
@Joxit I haven't, I was waiting for some clarifications in the other issue. Sure, go ahead and take it, thanks a ton for doing this! :) 🥂

**Comment by Joxit** [comment]
It seems that in order to validate this issue, all the importers must support the multi-lang index.

At this time, only [OSM](https://github.com/pelias/openstreetmap/blob/039513cae54c377ced75104e53a325e96b1e602a/stream/tag_mapper.js#L39) supports it.
WOF will be supported with pelias/whosonfirst#446
Geonames needs `alternateNamesV2` file to add multi-lang (we want that ?)
OpenAddresses and Polylines are unavailable

I think, the most important importer is WOF, the city/country search is the most common use case of the geocoder.

**Comment by mihneadb** [comment]
@Joxit FWIW there is some level of support for that already - when looking for something, passing e.g. `lang=en` or `lang=ru` yields the same `name` but the city name is translated.

https://pelias.github.io/compare/#/v1/autocomplete%3Flang=en&text=red%20square%20moscow
vs
https://pelias.github.io/compare/#/v1/autocomplete%3Flang=ru&text=red%20square%20moscow
(see `label`)

I thought that data was based on WOF.

**Comment by Joxit** [comment]
Yes, this is done by [pelias/placeholder](https://github.com/pelias/placeholder) which is a middleware and translate ElasticSearch responses for the user (using WOF ids).
This issue is about ES requests (and not responses).
That means, when you use `lang=ru` and search [red square Москва](https://pelias.github.io/compare/#/v1/autocomplete%3Flang=ru&text=red%20square%20%D0%9C%D0%BE%D1%81%D0%BA%D0%B2%D0%B0), you will not found the correct venue (`geonames:venue:6295575`).

The data is present in WOF, but not indexed in ES, only the default name and English variant are currently indexed. That's why I opened pelias/whosonfirst#446 :smile:

**Comment by mihneadb** [comment]
Gotcha now, thanks! About that, I'm thinking we should also return `Кра́сная пло́щадь` if someone searches `red square lang=ru`, would you agree? I'm thinking this should be easier to achieve - building on what you pointed out about the middleware. I can make a PR if so.

**Comment by Joxit** [comment]
I think the API can return the `name.{lang}` index when it's available in OSM, but for Geonames, it will be a bit more tricky because we do not use it anywhere.
Maybe this can be added in placeholder ? But we will have conflicts with WOF data...

**Comment by mihneadb** [comment]
I was thinking about it at a higher level. Simplest seems to me to update `geojsonify` here: https://github.com/pelias/api/blob/master/helper/geojsonify.js#L55-L60

Instead of going for `default`, prioritize `req.lang`?

**Comment by mihneadb** [comment]
Hi everyone! Any update on this one? LMK if I can help some way.

**Comment by missinglink** [comment]
Hi @mihneadb, unfortunately, it's me whos the blocker here, I would like to land https://github.com/pelias/api/pull/1287 before merging this (It's a complex change but I'm planning on doing the final testing and merging next week).

It's really not ideal to hold back another PR, especially a community contribution, but it makes sense for us in this case because the PR I linked is a massive refactoring of how autocomplete queries are generated.

We are sometimes a little over-cautious with merging big PRs but it's our responsibility to ensure compatibility and reliability for organisations running Pelias in a production environment with user-facing traffic.

**Comment by missinglink** [comment]
Oh actually I thought this was another PR, but the same still applies to this one ;)

**Comment by mihneadb** [comment]
@missinglink thanks for the transparency! Looking forward to using the new parser! :)

**Comment by slvlirnoff** [comment]
@missinglink Any news on this?

**Comment by missinglink** [comment]
I've been sick this week but releasing the new parser is a top priority.

**Comment by bboure** [comment]
Hi,

I came across an issue related to this today.  I was looking for [Edo Tokyo Museum](https://pelias.github.io/compare/#/v1/autocomplete?layers=venue&debug=1&text=Edo+Tokyo+Museum) and could not find any result. I realized that I had to search for [江戸東京博物館](https://pelias.github.io/compare/#/v1/autocomplete?layers=venue&text=%E6%B1%9F%E6%88%B8%E6%9D%B1%E4%BA%AC%E5%8D%9A%E7%89%A9%E9%A4%A8&debug=1) in order to find it.

Any ETA for this feature?

**Comment by Joxit** [comment]
Hi @bboure, this part of the feature is already live if you are running your query with `lang=en`. I found a difference in ES query between the English version and the Kanji version.

```json
{
  "constant_score": {
    "filter": {
      "multi_match": {
        "type": "cross_fields",
        "query": "Museum",
        "fields": [
          "parent.country.ngram^1",
          "parent.dependency.ngram^1",
          "parent.macroregion.ngram^1",
          "parent.region.ngram^1",
          "parent.county.ngram^1",
          "parent.localadmin.ngram^1",
          "parent.locality.ngram^1",
          "parent.borough.ngram^1",
          "parent.neighbourhood.ngram^1",
          "parent.locality_a.ngram^1",
          "parent.region_a.ngram^1",
          "parent.country_a.ngram^1",
          "name.default^1.5"
        ],
        "analyzer": "peliasQuery"
      }
    }
  }
}
```

In the `must` clause, `name.en^1.5` is missing.

The missing feature is multi lang in parent hierarchy now.

**Comment by bboure** [comment]
@Joxit Thanks for reaching back.

Add `lang=en` does not work either though. The query does not include `name.en^1.5`

https://pelias.github.io/compare/#/v1/autocomplete?layers=venue&text=Edo+Tokyo+Museum&debug=1

Am I doing something wrong?

**Comment by bboure** [comment]
Interestingly, looking for `Edo Tokyo Museum, Tokyo` [works](https://pelias.github.io/compare/#/v1/autocomplete?lang=en&text=edo+Tokyo+Museum%2C+Tokyo&debug=1)

It has to do on how the query is built

Edo Tokyo Museum, Tokyo:
````
"must": [
                   {
                      "multi_match": {
                        "type": "phrase",
                        "query": "edo Tokyo Museum",
                        "fields": [
                          "phrase.default",
                          "phrase.en"
                        ],
                        "analyzer": "peliasQuery",
                        "boost": 1,
                        "slop": 3
                      }
                    },
                   {
                      "multi_match": {
                        "type": "cross_fields",
                        "query": "Tokyo",
                        "fields": [
                          "parent.country.ngram^1",
                          "parent.dependency.ngram^1",
                          "parent.macroregion.ngram^1",
                          "parent.region.ngram^1",
                          "parent.county.ngram^1",
                          "parent.localadmin.ngram^1",
                          "parent.locality.ngram^1",
                          "parent.borough.ngram^1",
                          "parent.neighbourhood.ngram^1",
                          "parent.locality_a.ngram^1",
                          "parent.region_a.ngram^1",
                          "parent.country_a.ngram^1",
                          "name.default^1.5"
                        ],
                        "analyzer": "peliasAdmin"
                      }
                    }
                  ],
````
The full text falls into the `peliasQuery` analyzer here, and `Tokyo` into `peliasAdmin`

Edo Tokyo Museum:
```
"must": [
                   {
                      "multi_match": {
                        "type": "phrase",
                        "query": "edo Tokyo",
                        "fields": [
                          "phrase.default",
                          "phrase.en"
                        ],
                        "analyzer": "peliasQuery",
                        "boost": 1,
                        "slop": 3
                      }
                    },
                   {
                      "constant_score": {
                        "filter": {
                          "multi_match": {
                            "type": "cross_fields",
                            "query": "Museum",
                            "fields": [
                              "parent.country.ngram^1",
                              "parent.dependency.ngram^1",
                              "parent.macroregion.ngram^1",
                              "parent.region.ngram^1",
                              "parent.county.ngram^1",
                              "parent.localadmin.ngram^1",
                              "parent.locality.ngram^1",
                              "parent.borough.ngram^1",
                              "parent.neighbourhood.ngram^1",
                              "parent.locality_a.ngram^1",
                              "parent.region_a.ngram^1",
                              "parent.country_a.ngram^1",
                              "name.default^1.5"
                            ],
                            "analyzer": "peliasQuery"
                          }
                        }
                      }
],
```
`Museum` here is separated into a second rule and missing `name.en^1.5`

**Comment by Joxit** [comment]
Don't worry, I'm working on a fix, I will publish something tonight or tomorrow.

Yes, in autocomplete, the last token can be either a part of the subject (the venue) or the hierarchy. That's why we are using a `cross_fields` with both parent.* and name.default.

**Comment by bboure** [comment]
Great, thanks!
